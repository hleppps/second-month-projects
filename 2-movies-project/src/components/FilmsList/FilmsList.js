import React, { useContext } from "react";
import { Col, Row } from "react-bootstrap";
import AuthContext from "../../store/auth-context";
import { setFavoriteFilm } from "../../utils/api";
import FilmsListItem from "../FilmsListItem";
import LoadingSpinner from "../LoadingSpinner";
import styles from "./FilmsList.module.scss";

const FilmsList = React.memo((props) => {
  const authCtx = useContext(AuthContext);
  let filmsList;

  const addFavoriteFilmHandler = (filmId) => {
    authCtx.setFavoriteFilms((prevState) => [...prevState, filmId]);
    setFavoriteFilm(true, filmId, authCtx.token);
  };

  const removeFavoriteFilmHandler = (filmId) => {
    authCtx.setFavoriteFilms((prevState) =>
      prevState.filter((itemId) => itemId !== filmId)
    );
    setFavoriteFilm(false, filmId, authCtx.token);
  };

  const favoriteFilmHandler = (filmId, isFavorite) => {
    if (isFavorite) {
      removeFavoriteFilmHandler(filmId);
    } else {
      addFavoriteFilmHandler(filmId);
    }
  };

  if (props.status === "completed" && !!props.films.length) {
    filmsList = (
      <Row xs={1} md={2} xl={4} className="g-4">
        {props.films.map((filmItem) => (
          <Col className="g-4" key={filmItem.id}>
            <FilmsListItem
              setFavoriteFilm={favoriteFilmHandler}
              isFavorite={authCtx.favoriteFilms.includes(filmItem.id)}
              filmId={filmItem.id}
              backdropPath={filmItem.backdrop_path}
              title={filmItem.title}
              rating={filmItem.vote_average}
            />
          </Col>
        ))}
      </Row>
    );
  } else if (props.status === "completed" && !props.films.length) {
    filmsList = <p>No movies found by these search criteria...</p>;
  } else {
    filmsList = <LoadingSpinner />;
  }

  return (
    <div data-testid="films-list" className={styles.filmsList}>
      {filmsList}
    </div>
  );
});

export default FilmsList;
