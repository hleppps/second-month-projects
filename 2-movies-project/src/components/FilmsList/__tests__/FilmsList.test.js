import { cleanup, render } from "@testing-library/react";
import React from "react";
import FilmsList from "../FilmsList";

afterEach(cleanup);

describe("FilmsList components", () => {
  it("renders with content", () => {
    const { getByTestId } = render(<FilmsList />);

    expect(getByTestId("films-list")).not.toBeNull();
  });

  it("matches a snapshot", () => {
    const { asFragment } = render(<FilmsList />);
    expect(asFragment(<FilmsList />)).toMatchSnapshot();
  });
});
