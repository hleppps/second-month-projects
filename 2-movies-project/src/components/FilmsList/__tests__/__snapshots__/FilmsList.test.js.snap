// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmsList components matches a snapshot 1`] = `
<DocumentFragment>
  <div
    class="filmsList"
    data-testid="films-list"
  >
    <div
      class="spinner"
    />
  </div>
</DocumentFragment>
`;
